did_doc_builder/src/schema/did_doc.rs

```diff
@@ -0,0 +1,349 @@
+use std::collections::HashMap;
+
+use did_parser::{ParsedDID, ParsedDIDUrl};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+use super::{
+    service::Service,
+    types::uri::Uri,
+    utils::OneOrList,
+    verification_method::{VerificationMethod, VerificationMethodAlias},
+};
+
+type ControllerAlias = OneOrList<ParsedDID>;
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Default)]
+#[serde(default)]
+#[serde(rename_all = "camelCase")]
+pub struct DIDDocument {
+    id: ParsedDID,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    also_known_as: Vec<Uri>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    controller: Option<ControllerAlias>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    verification_method: Vec<VerificationMethod>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    authentication: Vec<VerificationMethodAlias>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    assertion_method: Vec<VerificationMethodAlias>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    key_agreement: Vec<VerificationMethodAlias>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    capability_invocation: Vec<VerificationMethodAlias>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    capability_delegation: Vec<VerificationMethodAlias>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    service: Vec<Service>,
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    #[serde(flatten)]
+    extra: HashMap<String, Value>,
+}
+
+impl DIDDocument {
+    pub fn builder(id: ParsedDID) -> DIDDocumentBuilder {
+        DIDDocumentBuilder::new(id)
+    }
+
+    pub fn id(&self) -> &ParsedDID {
+        &self.id
+    }
+
+    pub fn also_known_as(&self) -> &[Uri] {
+        self.also_known_as.as_ref()
+    }
+
+    pub fn controller(&self) -> Option<&OneOrList<ParsedDID>> {
+        self.controller.as_ref()
+    }
+
+    pub fn verification_method(&self) -> &[VerificationMethod] {
+        self.verification_method.as_ref()
+    }
+
+    pub fn authentication(&self) -> &[VerificationMethodAlias] {
+        self.authentication.as_ref()
+    }
+
+    pub fn assertion_method(&self) -> &[VerificationMethodAlias] {
+        self.assertion_method.as_ref()
+    }
+
+    pub fn key_agreement(&self) -> &[VerificationMethodAlias] {
+        self.key_agreement.as_ref()
+    }
+
+    pub fn capability_invocation(&self) -> &[VerificationMethodAlias] {
+        self.capability_invocation.as_ref()
+    }
+
+    pub fn capability_delegation(&self) -> &[VerificationMethodAlias] {
+        self.capability_delegation.as_ref()
+    }
+
+    pub fn service(&self) -> &[Service] {
+        self.service.as_ref()
+    }
+
+    pub fn extra(&self, key: &str) -> Option<&Value> {
+        self.extra.get(key)
+    }
```

Comment:
Left in as is as discussed.

Reply:
How about renaming it then? The method looks like a getter of the map field instead of a lookup into the map. I agree though that the inner field should not be exposed.

Reply:
Apologies, forgot about the suggestion to rename. What name do you suggest?

Reply:
Something like `lookup_extra_field`?

Reply:
How about `get_extra_field` instead of `lookup_extra_field`?

Reply:
I think get is implicit. Maybe just `extra_field`?